File: data-structures/Hash-Table/JS/hash-table.js
class Node {
  constructor(key, value, next = null) {
    this.key = key;
    this.value = value;
    this.next = next;
  }
}

class HashTable {
  #table;
  #capacity;
  #size;
  #loadFactor;

  constructor(capacity = 17, loadFactor = 1.0) {
    if (typeof capacity != 'number' || capacity <= 0) {
      throw new Error('Capacity must be positive number');
    }

    this.#table = new Array(capacity).fill(null);
    this.#size = 0;
    this.#capacity = capacity;
    this.#loadFactor = loadFactor;
  }

  /* ================= Basic State ================= */

  size() {
    return this.#size;
  }

  capacity() {
    return this.#capacity;
  }

  isEmpty() {
    return this.#size === 0;
  }

  clear() {
    this.#size = 0;
    this.#table = new Array(this.#capacity).fill(null);
  }

  /* ================= Hashing ================= */

  #hash(key) {
    //Division Method
    if (typeof key === 'number') {
      return Math.floor(Math.abs(key) % this.#capacity);
    }

    // Polynomial Hash
    if (typeof key === 'string') {
      let h = 0;
      const p = 31;

      for (let i = 0; i < key.length; i++) {
        h = (h * p + key.charCodeAt(i)) | 0;
      }

      return Math.abs(h) % this.#capacity;
    }
    throw new Error('Unsupported key type');
  }

  /* ================= Core Operations ================= */

  put(key, value) {
    let index = this.#hash(key);
    let node = this.#table[index];

    while (node) {
      if (node.key === key) {
        node.value = value;
        return;
      }
      node = node.next;
    }

    const newNode = new Node(key, value);
    newNode.next = this.#table[index];
    this.#table[index] = newNode;
    ++this.#size;

    if (this.loadFactor() > this.#loadFactor) {
      this.#resize(this.#capacity * 2);
    }
    ++this.#size;
    return;
  }

  get(key) {
    let index = this.#hash(key);
    let node = this.#table[index];

    while (node) {
      if (node.key === key) {
        return node.value;
      }
    }
    return undefined;
  }

  remove(key) {
    let index = this.#hash(key);
    let node = this.#table[index];
    let prev = null;
    let removed;

    while (node) {
      if (node.key === key) {
        removed = node.value;
        if (prev) {
          prev.next = node.next;
          node.next = null;
          --this.#size;
          return removed;
        }
        this.#table[index] = node.next;
        --this.#size;
        return removed;
      }
      prev = node;
      node = node.next;
    }
  }

  containsKey(key) {
    let index = this.#hash(key);
    let node = this.#table[index];

    while (node) {
      if (node.key === key) {
        return true;
      }
      node = node.next;
    }
    return false;
  }

  containsValue(value) {
    for (let i = 0; i < this.#capacity; ++i) {
      let node = null;
      if (this.#table[i]) {
        node = this.#table[i];
      } else {
        continue;
      }

      while (node) {
        if (node.value === value) {
          return true;
        }
        node = node.next;
      }
    }
    return false;
  }
  /* ================= Resize / Rehash ================= */

  #resize(newCapacity) {
    const prime = this.#getNextPrime(newCapacity);
    const newBuckets = new Array(prime).fill(null);
    const oldTable = this.#table;

    this.#table = newBuckets;
    this.#capacity = prime;

    for (let head of oldTable) {
      let node = head;

      while (node) {
        this.put(node.key, node.value);
        node = node.next;
      }
    }
  }

  loadFactor() {
    return this.#size / this.#capacity;
  }

  /* ================= isPrime / getNextPrime ================= */

  #isPrime(num) {
    if (num <= 1) {
      return false;
    }

    if (num === 2) {
      return true;
    }

    if (num % 2 === 0) {
      return false;
    }

    const sqrtNum = Math.floor(Math.sqrt(num));
    for (let i = 3; i <= sqrtNum; i += 2) {
      if (num % i === 0) {
        return false;
      }
    }
    return true;
  }

  #getNextPrime(num) {
    let nextPrime = num + 1;
    while (!this.#isPrime(nextPrime)) {
      nextPrime++;
    }
    return nextPrime;
  }

  /* ================= Entry Views ================= */

  keys() {
    const keys = [];

    for (let i = 0; i < this.#capacity; ++i) {
      let node = null;

      if (this.#table[i]) {
        node = this.#table[i];
      } else {
        continue;
      }

      while (node) {
        keys.push(node.key);
        node = node.next;
      }
    }
    return keys;
  }

  values() {
    const values = [];

    for (let i = 0; i < this.#capacity; ++i) {
      let node = null;

      if (this.#table[i]) {
        node = this.#table[i];
      } else {
        continue;
      }

      while (node) {
        values.push(node.value);
        node = node.next;
      }
    }
    return values;
  }

  entries() {
    const entries = [];

    for (let i = 0; i < this.#capacity; ++i) {
      let node = null;

      if (this.#table[i]) {
        node = this.#table[i];
      } else {
        continue;
      }

      while (node) {
        entries.push([node.key, node.value]);
        node = node.next;
      }
    }
    return entries;
  }

  /* ================= Iteration ================= */

  [Symbol.iterator]() {
    const entries = this.entries();
    let i = 0;

    return {
      next: () => {
        if (i < entries.length) {
          return {
            value: entries[i++],
            done: false,
          };
        }
        return {
          value: undefined,
          done: true,
        };
      },
    };
  }

  /* ================= Utility Operations ================= */

  toObject() {
    const map = new Map();
    const entries = this.entries();

    for (let i = 0; i < entries.length; ++i) {
      map.set(entries[i][0], entries[i][1]);
    }
    return map;
  }

  clone() {
    const clone = new HashTable(this.#capacity, this.#loadFactor);

    for (let head of this.#table) {
      let node = head;

      while (node) {
        clone.put(node.key, node.value);
        node = node.next;
      }
    }
    return clone;
  }

  equals(otherTable) {
    if (
      otherTable.size() !== this.size() ||
      !(otherTable instanceof HashTable) ||
      otherTable.capacity() !== this.capacity()
    ) {
      return false;
    }

    const thisEntries = this.entries();
    const otherEntries = otherTable.entries();

    for (let i = 0; i < thisEntries.length; ++i) {
      if (
        thisEntries[i][0] !== otherEntries[i][0] ||
        thisEntries[i][1] !== otherEntries[i][1]
      ) {
        return false;
      }
    }
    return true;
  }

  /* ================= Debug / Visualization ================= */

  bucketSizes() {
    const sizes = [];

    for (let i = 0; i < this.#capacity; ++i) {
      let count = 0;
      let node = null;

      if (this.#table[i]) {
        node = this.#table[i];
      } else {
        sizes.push(0);
        continue;
      }

      while (node) {
        ++count;
        node = node.next;
      }
      sizes.push(count);
    }
    return sizes;
  }

  print() {
    for (let i = 0; i < this.#capacity; i++) {
      let node = this.#table[i];
      let output = `Bucket ${i}: `;

      while (node) {
        output += `(${node.key}: ${node.value}) -> `;
        node = node.next;
      }

      console.log(output + 'null');
    }
  }
}

File: data-structures/Hash-Table/C/src/hash_table.c
#include "../include/hash_table.h"

#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* =======================
   Static Helpers
   ======================= */

bool is_prime(size_t num)
{
    for (size_t i = 2; i * i <= num; ++i)
    {
        if (num % i == 0)
        {
            return false;
        }
    }

    return true;
}

size_t get_next_prime(size_t num)
{
    while (!is_prime(num))
    {
        ++num;
    }

    return num;
}

/* =======================
   Core Operations
   ======================= */

/* Create & Destroy */
HashTable* ht_create(size_t capacity)
{
    if (!is_prime(capacity))
    {
        capacity = get_next_prime(capacity);
    }

    HashTable* ht = (HashTable*)malloc(sizeof(HashTable));
    if (!ht)
    {
        free(ht);
        return NULL;
    }

    ht->buckets = (HashNode**)malloc(sizeof(HashNode*) * capacity);
    if (!ht->buckets)
    {
        return NULL;
    }

    for (size_t i = 0; i < capacity; ++i)
    {
        ht->buckets[i] = NULL;
    }

    ht->capacity = capacity;
    ht->size = 0;
    ht->max_load_factor = 0.75;

    return ht;
}

void ht_destroy(HashTable* ht)
{
    if (!ht)
    {
        return;
    }

    for (size_t i = 0; i < ht->capacity; i++)
    {
        if (ht->buckets[i] != NULL)
        {
            HashNode* current = ht->buckets[i];

            while (current)
            {
                HashNode* tmp = current;
                current = current->next;

                free(tmp->key);
                free(tmp);
            }
        }
    }

    free(ht->buckets);
    free(ht);
}

/* Insert or Update */
HT_ErrorCode ht_put(HashTable* ht, const char* key, int value)
{
    if (!ht || !key)
    {
        return HT_ERR_NULL_POINTER;
    }

    if ((double)(ht->size + 1) / ht->capacity > ht->max_load_factor)
    {
        HT_ErrorCode err = ht_resize(ht, ht->capacity * 2);
        if (err != HT_OK)
        {
            return err;
        }
    }

    size_t index = ht_hash(key, ht->capacity);
    HashNode* current = ht->buckets[index];

    while (current)
    {
        if (strcmp(current->key, key) == 0)
        {
            current->value = value;
            return HT_OK;
        }
        current = current->next;
    }

    HashNode* newNode = (HashNode*)malloc(sizeof(HashNode));
    if (!newNode)
    {
        return HT_ERR_OUT_OF_MEMORY;
    }
    newNode->key = strdup(key);
    if (!newNode->key)
    {
        free(newNode);
        return HT_ERR_OUT_OF_MEMORY;
    }
    newNode->value = value;

    newNode->next = ht->buckets[index];
    ht->buckets[index] = newNode;
    ++ht->size;

    return HT_OK;
}

/* Get value by key */
HT_ErrorCode ht_get(HashTable* ht, const char* key, int* out_value)
{
    if (!ht || !key || !out_value)
    {
        return HT_ERR_NULL_POINTER;
    }

    size_t index = ht_hash(key, ht->capacity);
    HashNode* current = ht->buckets[index];

    while (current)
    {
        if (strcmp(current->key, key) == 0)
        {
            *out_value = current->value;
            return HT_OK;
        }
        current = current->next;
    }
    return HT_ERR_KEY_NOT_FOUND;
}

/* Remove key */
HT_ErrorCode ht_remove(HashTable* ht, const char* key)
{
    if (!ht || !key)
    {
        return HT_ERR_NULL_POINTER;
    }

    size_t index = ht_hash(key, ht->capacity);
    HashNode* current = ht->buckets[index];
    HashNode* prev = NULL;

    while (current)
    {
        if (strcmp(current->key, key) == 0)
        {
            if (!prev)
            {
                ht->buckets[index] = current->next;
            }
            else
            {
                prev->next = current->next;
                break;
            }
            free(current->key);
            free(current);
            --ht->size;
            return HT_OK;
        }
        prev = current;
        current = current->next;
    }

    return HT_ERR_KEY_NOT_FOUND;
}

/* Check existence */
bool ht_contains(HashTable* ht, const char* key)
{
    if (!ht || !key)
    {
        return false;
    }

    size_t index = ht_hash(key, ht->capacity);
    HashNode* current = ht->buckets[index];

    while (current)
    {
        if (strcmp(current->key, key) == 0)
        {
            return true;
        }
        current = current->next;
    }
    return false;
}

/* Size */
size_t ht_size(HashTable* ht)
{
    if (!ht)
    {
        return 0;
    }

    return ht->size;
}

/* Hash function */
size_t ht_hash(const char* key, size_t capacity)
{
    if (!key || capacity == 0)
    {
        return 0;
    }

    unsigned long hash = 0;
    const unsigned long p = 31;

    while (*key)
    {
        hash = hash * p + (unsigned char)(*key);
        key++;
    }

    return hash % capacity;
}

/*Check Load Factor*/
double ht_load_factor(HashTable* ht)
{
    if (!ht)
    {
        return 0.0;
    }

    return (double)ht->size / ht->capacity;
}

/*Print Hash Table*/
void ht_print(HashTable* ht)
{
    if (!ht)
    {
        printf("HashTable is NULL\n");
        return;
    }

    for (size_t i = 0; i < ht->capacity; i++)
    {
        HashNode* node = ht->buckets[i];

        printf("Bucket [%zu]: ", i);

        while (node)
        {
            printf("(%s: %d) -> ", node->key, node->value);
            node = node->next;
        }

        printf("null\n");
    }
}

/* =======================
   Resize / Rehash
   ======================= */

/* Resize table to new capacity */
HT_ErrorCode ht_resize(HashTable* ht, size_t new_capacity)
{
    if (!ht)
    {
        return HT_ERR_NULL_POINTER;
    }

    new_capacity = get_next_prime(new_capacity);

    HashNode** new_buckets =
        (HashNode**)malloc(sizeof(HashNode*) * new_capacity);
    if (!new_buckets)
    {
        return HT_ERR_OUT_OF_MEMORY;
    }

    for (size_t i = 0; i < new_capacity; ++i)
    {
        new_buckets[i] = NULL;
    }

    for (size_t i = 0; i < ht->capacity; ++i)
    {
        HashNode* current = ht->buckets[i];

        while (current)
        {
            HashNode* next = current->next;

            size_t new_index = ht_hash(current->key, new_capacity);

            current->next = new_buckets[new_index];
            new_buckets[new_index] = current;

            current = next;
        }
    }

    free(ht->buckets);

    ht->buckets = new_buckets;
    ht->capacity = new_capacity;

    return HT_OK;
}